src/txorphanage.h

```diff
@@ -21,65 +21,76 @@ extern RecursiveMutex g_cs_orphans;
+     *  peer, and whether there are more orphans for this peer to work on
+     *  after this tx.
      */
-    std::pair<CTransactionRef, NodeId> GetTx(const uint256& txid) const EXCLUSIVE_LOCKS_REQUIRED(g_cs_orphans);
+    bool GetTxToReconsider(NodeId peer, CTransactionRef& ref, NodeId& originator, bool& more) EXCLUSIVE_LOCKS_REQUIRED(!m_mutex);
 
     /** Erase an orphan by txid */
-    int EraseTx(const uint256& txid) EXCLUSIVE_LOCKS_REQUIRED(g_cs_orphans);
+    int EraseTx(const uint256& txid) EXCLUSIVE_LOCKS_REQUIRED(!m_mutex) { LOCK(m_mutex); return _EraseTx(txid); }
 
     /** Erase all orphans announced by a peer (eg, after that peer disconnects) */
-    void EraseForPeer(NodeId peer) EXCLUSIVE_LOCKS_REQUIRED(g_cs_orphans);
+    void EraseForPeer(NodeId peer) EXCLUSIVE_LOCKS_REQUIRED(!m_mutex);
 
     /** Erase all orphans included in or invalidated by a new block */
-    void EraseForBlock(const CBlock& block) LOCKS_EXCLUDED(::g_cs_orphans);
+    void EraseForBlock(const CBlock& block) EXCLUSIVE_LOCKS_REQUIRED(!m_mutex);
 
     /** Limit the orphanage to the given maximum */
-    unsigned int LimitOrphans(unsigned int max_orphans) EXCLUSIVE_LOCKS_REQUIRED(g_cs_orphans);
+    unsigned int LimitOrphans(unsigned int max_orphans) EXCLUSIVE_LOCKS_REQUIRED(!m_mutex);
 
-    /** Add any orphans that list a particular tx as a parent into a peer's work set
-     * (ie orphans that may have found their final missing parent, and so should be reconsidered for the mempool) */
-    void AddChildrenToWorkSet(const CTransaction& tx, std::set<uint256>& orphan_work_set) const EXCLUSIVE_LOCKS_REQUIRED(g_cs_orphans);
+    /** Add any orphans that list a particular tx as a parent into a peer's work set */
+    void AddChildrenToWorkSet(const CTransaction& tx, NodeId peer) EXCLUSIVE_LOCKS_REQUIRED(!m_mutex);
 
 protected:
+    /** Guards data */
+    mutable Mutex m_mutex;
+
+    /** Erase an orphan by txid (internal, lock must already be held) */
+    int _EraseTx(const uint256& txid) EXCLUSIVE_LOCKS_REQUIRED(m_mutex);
+
     struct OrphanTx {
         CTransactionRef tx;
         NodeId fromPeer;
         int64_t nTimeExpire;
         size_t list_pos;
     };
 
-    /** Map from txid to orphan transaction record. Limited by
-     *  -maxorphantx/DEFAULT_MAX_ORPHAN_TRANSACTIONS */
-    std::map<uint256, OrphanTx> m_orphans GUARDED_BY(g_cs_orphans);
-
-    using OrphanMap = decltype(m_orphans);
+    using OrphanMap = std::map<uint256, OrphanTx>;
 
     struct IteratorComparator
     {
-        template<typename I>
-        bool operator()(const I& a, const I& b) const
+        bool operator()(const OrphanMap::iterator& a, const OrphanMap::iterator& b) const
         {
             return &(*a) < &(*b);
         }
     };
 
+    /** Map from txid to orphan transaction record. Limited by
+     *  -maxorphantx/DEFAULT_MAX_ORPHAN_TRANSACTIONS */
+    OrphanMap m_orphans GUARDED_BY(m_mutex);
```

Comment:
Is it possible for `TxOrphanage::m_orphans` to contain keys (transaction ids) which are not in `TxOrphanage::m_peer_work_set` or the other way around?

Reply:
`m_orphans` will normally contain txids that aren't in `m_peer_work_set` -- they only get added to the work set when a parent appears and are (hopefully) very quickly removed from the work set after retrying ATMP.

It's also possible for something to expire from `m_orphans` and still be present in `m_peer_work_set`: if a tx that's in the work set is removed via `LimitOrphans` or `EraseForBlock` you'll get that case. It will eventually be removed from the work set when `GetTxToReconsider` is called.